examples/hal/quad/main.rs

```diff
@@ -3,614 +3,1380 @@
-extern crate gfx_backend_vulkan as back;
-#[cfg(feature = "metal")]
-extern crate gfx_backend_metal as back;
 #[cfg(feature = "gl")]
 extern crate gfx_backend_gl as back;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "vulkan")]
+extern crate gfx_backend_vulkan as back;
+#[cfg(not(any(feature = "vulkan", feature = "dx12", feature = "metal", feature = "gl")))]
+extern crate gfx_backend_empty as back;
 
-extern crate winit;
+extern crate env_logger;
+extern crate gfx_hal as hal;
+extern crate glsl_to_spirv;
 extern crate image;
+extern crate takeable_option;
+extern crate winit;
+
+struct Dimensions<T> {
+    width: T,
+    height: T,
+}
+
+use std::cell::RefCell;
+use std::io::Cursor;
+use std::mem::size_of;
+use std::mem::ManuallyDrop;
+use std::rc::Rc;
 
-use hal::{buffer, command, format as f, image as i, memory as m, pass, pso, pool};
-use hal::{Device, Instance, PhysicalDevice, Surface, Swapchain};
 use hal::{
-    DescriptorPool, FrameSync, Primitive,
-    Backbuffer, SwapchainConfig,
+    buffer, command, format as f, image as i, memory as m, pass, pool, pso, Adapter, Backbuffer,
+    Backend, DescriptorPool, Device, FrameSync, Instance, Limits, MemoryType, PhysicalDevice,
+    Primitive, QueueGroup, Surface, Swapchain, SwapchainConfig,
 };
+
 use hal::format::{AsFormat, ChannelType, Rgba8Srgb as ColorFormat, Swizzle};
 use hal::pass::Subpass;
 use hal::pso::{PipelineStage, ShaderStageFlags, Specialization};
 use hal::queue::Submission;
+use takeable_option::Takeable;
 
-use std::io::Cursor;
+use std::fs::File;
+use std::io::Read;
 
 const ENTRY_NAME: &str = "main";
 
 #[derive(Debug, Clone, Copy)]
-#[allow(non_snake_case)]
 struct Vertex {
-    a_Pos: [f32; 2],
-    a_Uv: [f32; 2],
+    a_pos: [f32; 2],
+    a_uv: [f32; 2],
 }
 
 const QUAD: [Vertex; 6] = [
-    Vertex { a_Pos: [ -0.5, 0.33 ], a_Uv: [0.0, 1.0] },
-    Vertex { a_Pos: [  0.5, 0.33 ], a_Uv: [1.0, 1.0] },
-    Vertex { a_Pos: [  0.5,-0.33 ], a_Uv: [1.0, 0.0] },
-
-    Vertex { a_Pos: [ -0.5, 0.33 ], a_Uv: [0.0, 1.0] },
-    Vertex { a_Pos: [  0.5,-0.33 ], a_Uv: [1.0, 0.0] },
-    Vertex { a_Pos: [ -0.5,-0.33 ], a_Uv: [0.0, 0.0] },
+    Vertex {
+        a_pos: [-0.5, 0.33],
+        a_uv: [0.0, 1.0],
+    },
+    Vertex {
+        a_pos: [0.5, 0.33],
+        a_uv: [1.0, 1.0],
+    },
+    Vertex {
+        a_pos: [0.5, -0.33],
+        a_uv: [1.0, 0.0],
+    },
+    Vertex {
+        a_pos: [-0.5, 0.33],
+        a_uv: [0.0, 1.0],
+    },
+    Vertex {
+        a_pos: [0.5, -0.33],
+        a_uv: [1.0, 0.0],
+    },
+    Vertex {
+        a_pos: [-0.5, -0.33],
+        a_uv: [0.0, 0.0],
+    },
 ];
 
 const COLOR_RANGE: i::SubresourceRange = i::SubresourceRange {
     aspects: f::Aspects::COLOR,
-    levels: 0 .. 1,
-    layers: 0 .. 1,
+    levels: 0..1,
+    layers: 0..1,
 };
 
-#[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal", feature = "gl"))]
-fn main() {
-    env_logger::init();
+struct RendererState<B: Backend> {
```

Comment:
ah, ok. That's definitely a bit more complex than we'd like for this example. Maybe for another one, a more general/complex one, we'll need a struct, but this one is supposed to be as simple as possible.